src/kili/graphql/__init__.py

```diff
@@ -149,43 +145,36 @@ def execute_query_from_paginated_call(
                     count_rows_retrieved += len(rows)
                     pbar.update(len(rows))
 
-    def fragment_builder(self, fields: List[str], typed_dict_class: Type[TypedDict]):
+    @typechecked
+    def fragment_builder(self, fields: List[str]):
         """
         Builds a GraphQL fragment for a list of fields to query
 
         Args:
-            fields
-            type_of_fields
+            fields: The list of fields to query
         """
-        type_of_fields = typed_dict_class.__annotations__
         fragment = ""
+
+        # split a field and its subfields (e.g. "roles.user.id" -> ["roles", "user.id"])
         subfields = [field.split(".", 1) for field in fields if "." in field]
+
         if subfields:
-            for subquery in {subfield[0] for subfield in subfields}:
-                type_of_fields_subquery = type_of_fields[subquery]
-                if type_of_fields_subquery == str:
-                    raise NonExistingFieldError(f"{subquery} field does not take subfields")
-                if is_typeddict(type_of_fields_subquery):
-                    fields_subquery = [
-                        subfield[1] for subfield in subfields if subfield[0] == subquery
-                    ]
-                    new_fragment = self.fragment_builder(
-                        fields_subquery,
-                        type_of_fields_subquery,  # type: ignore
-                    )
-                    fragment += f" {subquery}{{{new_fragment}}}"
+            # get the root fields (e.g. "roles" in "roles.user.id")
+            root_fields = {subfield[0] for subfield in subfields}
+            for root_field in root_fields:
+                # get the subfields of the root field (e.g. "user.id" in "roles.user.id")
+                fields_subquery = [
+                    subfield[1] for subfield in subfields if subfield[0] == root_field
+                ]
+                # build the subquery fragment (e.g. "user{id}" in "roles{user{id}}")
+                new_fragment = self.fragment_builder(fields_subquery)
+                # add the subquery to the fragment
+                fragment += f" {root_field}{{{new_fragment}}}"
+
+            # remove the fields that have been queried in subqueries (e.g. "roles.user.id")
             fields = [field for field in fields if "." not in field]
+
         for field in fields:
-            try:
-                type_of_fields[field]
-            except KeyError as exception:
-                raise NonExistingFieldError(
-                    f"Cannot query field {field} on object {typed_dict_class.__name__}. Admissible"
-                    " fields are: \n- "
-                    + "\n- ".join(type_of_fields.keys())
-                ) from exception
-            if isinstance(field, str):
-                fragment += f" {field}"
-            else:
-                raise TypeError("Please provide the fields to query as strings")
```

Comment:
So all this feature of rejecting wrong fields is done by gql now? What message do I get if I ask for asset.do_not_exist for example ?

Reply:
yes, the query will be refused if the field is invalid

```text
graphql.error.graphql_error.GraphQLError: Cannot query field 'sdfsdfsdfsdfsdfsdfs' on type 'Asset'.

GraphQL request:4:21
3 |             data: assets(where: $where, skip: $skip, first: $first) {
4 |                  id sdfsdfsdfsdfsdfsdfs
  |                     ^
5 |             }
```